VSE compat, Alpha Implant compat, revamp general compatibility

cvrunmin opened the pull request:
More compatibilities with other mods, and some revamps on GeneralCompatibility for brain template and genome template.

## New features (User)
1. When genome sequence and brain scanner is in storage, genome sequencing and brain scanning operation will always be shown in "add surgery" now. These operations will be disabled will disabled reason appended on the labels if invalid races or hediffs blocking such operation.
2. When Vanilla Skills Expanded is installed, brain scanner will also store pawn's expertise.

## New features (Developer)
1. Add `blockCreatingBrainTemplate` and `blockCreatingGenomeTemplate` for `HediffTemplateProperties`, allowing modders to block brain scanning or genome sequencing separately.
4. Add `includeSeverityInTemplate` for `HediffTemplateProperties`, allowing modders to request templates to record severity of hediff if severity matters.
5. Add `pregeneratedInCloning` for `HediffTemplateProperties`, allowing modders to request cloning vat to keep the pregenerated hediff from being flushed out, but such hediff should not be recorded in templates. For example, a mod that add a gene which will create a special organ. Such organ can be replaced with artificial part and the modder decides that that artificial organ should not be grown with the clone, but instead the natural organ from the gene.

## Fixes
1. Fix brain scanning template not saving backstories into game save.
2. Fix genome sequencing and brain scanning operations not showing up even when the corresponding item exists in storage,  when Alpha Implants is loaded before QEE.

## Other changes
1. Add some documentation for `HediffTemplateProperties` and `RaceExclusionProperties`.
2. Wrap checking code related to `GeneralCompatibility` into methods, moved into `GeneralCompatibility`. Now such checking will primarily use values in `HediffTemplateProperties`.

EDIT (22 Feb 2025): Added New features # 1, # 2. Added Fixes # 1.

cvrunmin committed:
revamp general compatibility

emipa606 commented:
Someone commented on broken compatibility with evolved organs-mod with that last changes. Could you look into that before we do any more changes?

cvrunmin committed:
fix .net complaining about sorting with LabelCap

cvrunmin commented:
I've tried playing with only QEE and evolved organs. Pawn are cloned smoothly, extra limbs and ograns installed are correctly brought forward to the clone. Indeed I found out that .NET complains about sorting with def.LabelCap when checking if template is stackable. This is fixed in the latest commit.
disclaimer: I have never played with evolved organs before, so I don't know what functions would related to the changes made in last pull request. Also the comment have disappeared so I doubt if there is actually a incompatibility we are yet to find out. The message is also very vague to be reproducible.

cvrunmin committed:
use AvailableReport for Brainscanning and GenomeSequencing instead, attempt fix Alpha Implant compat

have to say the incompatibility with Alpha Implant is very weird. Surgeries are missing only when QEE comes after Alpha Implant. Maybe Alpha Implant somehow make base game cached the surgery list?

cvrunmin committed:
VanillaSkillsExpanded Compatibility, fix backStory not saving

cvrunmin committed:
Translations via Deepl

cvrunmin committed:
ensure expertises is non null

cvrunmin commented:
UPDATE (22 Feb 2025): more updates are made to make this PR contains new features that are observable in player side. New features include:
1. "Brain Scanning" and "Genome Sequencing" surgeries will appear when the corresponding item presents, regardless the actual availability to operate such surgery. Instead, these surgeries will be shown disabled with the reason why they are disabled.
2. When Vanilla Skills Expanded is installed, Brain scanning will also record source pawn's expertises.
3. Fix brain scan template not saving backstories into game save.
4. Fix "brain Scanning" and "Genome Sequencing" surgeries not showing up when Alpha Implants is loaded before QEE.
    - Indeed, if QEE loads before Alpha Implants, these surgeries will show up. If someone has to be blame, I suspect them as in  [AlphaImplants:VFECore.AlphaImplants](https://github.com/juanosarg/AlphaImplants/blob/f665b2916f2a073ea4bfd53dd3dd89b52bd6343f/1.5/Source/AlphaImplants/AlphaImplants/RefreshCaches.cs), they cached all recipes right at startup. Nevertheless I still attempt to fix it by following what they did, to patch our surgeries to the recipes of things. Checked some of mods that add surgeries and no one did this before except Alpha Implants, so I only activate this patch when Alpha Implants presents.

emipa606 commented:
Very nice! All published now in 713665c
